[Site Isolation] Drag & Drop within remote frame is nonfunctional

jesxilin opened the pull request:
#### cfed61549b572670717859637909a48031feb133
<pre>
[Site Isolation] Drag &amp; Drop within remote frame is nonfunctional
<a href="https://bugs.webkit.org/show_bug.cgi?id=301963">https://bugs.webkit.org/show_bug.cgi?id=301963</a>
<a href="https://rdar.apple.com/160806781">rdar://160806781</a>

Reviewed by Aditya Keerthi and Alex Christensen.

Original patch by pascoej. Thanks to achristensen07 for suggesting this
new approach.

In certain site isolation cases, dragging and dropping on macOS/iOS causes
a navigation when dropping items instead of a drop event.
The default mode for drag and drop is to load, which would then
trigger a navigation. When dropping with the presence of remote frames,
there is no logic to support this functionality and the behavior defaults
to load. To fix this unexpected behavior, implement support
for drag and drop with remote frames. This includes finding the correct
destination frame, and propagating the drag and drop action to the target
frame.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add new files that supports DragEventTargetData and DragEventForwardingData.
The DragEventForwardingData struct contains the information for propagating the event
to the target frame.

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::performDragOperation):
* Source/WebCore/page/DragController.h:
Alter the original code to factor in the presence of remote frames.
If a remote frame (not the current frame) is the destination of the
drop event, return a DragEventTargetData that represents the remote frame ID. This
remote frame ID will be used as the new destination frame ID. Otherwise,
handle cases for edit and load accordingly using the current frame instead
of the main frame.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::contentFrameForNode):
(WebCore::EventHandler::updateDragAndDrop):
(WebCore::EventHandler::cancelDragAndDrop):
(WebCore::EventHandler::performDragAndDrop):
* Source/WebCore/page/EventHandler.h:
The return value in EventHandler::performDragAndDrop indicates if a drop event
was successfully handled or not.

Check if the subframeForTargetNode is a remote frame or a local frame.
If the subframe is a remote frame, return the targetFrameID.

If the current frame is the owner of the target frame for the drop event,
there are two cases: local and remote frame. If the target frame is a local frame,
simply call performDragAndDrop on the target frame. In this case, the target frame
is accessible. If the target frame is remote, the target frame is not accessible
in this context. Thus, populate DragEventTargetData targetFrameID to be used to
propagate to the remote frame.

Otherwise, if it&apos;s at the target, create DataTransfer::createForDrop and
return whether the drop event was successfully executed.

* Source/WebCore/platform/DragEventTargetData.h: Added.
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/DragEventForwardingData.h: Added.
* Source/WebKit/Shared/DragEventForwardingData.serialization.in: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Add new files for the DragEventTargetData and DragOperationResult variants.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::propagateDragAndDrop):
(WebKit::WebPageProxy::performDragOperation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
DragOperationResult contains either a boolean that indicates the equivalent of
the DragEventHandled in DragEventTargetData (whether the drag/drop event was successful),
or a DragEventForwardingData struct that neatly packages all the relevant information needed
for the WebPageProxy to forward the drag event to the correct frame. DragEventForwardingData
consists of the targetFrameID and sandbox extension handles needed.

WebPageProxy::performDragOperation utilizes a new model for drag and drop to include remote frames.
Previously, the code only accounted for the local main frame.

This is the current model:

1) WebPageProxy&apos;s performDragOperation is responsible for sending the initial
IPC call to WebPage::PerformDragOperation, in the context of the main frame.

2) A reply (completion handler) is received from WebPage::PerformDragOperation.
If DragEventTargetData is of DragEventHandled type, it&apos;ll indicate if the event succeeded. In this
case, return a boolean in the completion handler.
If the type is a Frame Identifier, that means there is a frame that needs to receive this
operation and hasn&apos;t yet. So, create a DragEventForwardingData that includes all the information needed
for the operation and return it in the completion handler.

3) In propagateDragAndDrop, send an IPC message to WebPage::PerformDragOperation
with the DragEventForwardingData&apos;s targetFrameID. If it is still not handled and there is another Frame
Identifier that is not the current frame ID passed in, a propagation must happen again.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::performDragControllerAction):
(WebKit::WebPage::performDragOperation):
(WebKit::WebPage::willPerformLoadDragDestinationAction):
(WebKit::WebPage::mayPerformUploadDragDestinationAction):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
DragEventTargetData used in WebPage consists of a DragEventHandled that represents if the drag event
was successful or not, and a Frame Identifier that represents the frameID of the
target. Also, change the sandboxing logic to use SandboxExtension handles instead. This is because
when WebPage::performDragOperation is called via IPC messages, now
there is no guaranteee that the extensions will be used in the current frame context.
As a result, there needs to be a way to retain the SandboxExtension handles, which
will later be utilized to create the extensions to be consumed for dropping, on the
actual target frame.

* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _tryToPerformDataInteraction:client:global:operation:]):
(-[WebView performDragOperation:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, DragAndDropWithoutNavigation)):
Create an API test to ensure a navigation does not occur when dragging and dropping.

Canonical link: <a href="https://commits.webkit.org/306143@main">https://commits.webkit.org/306143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a63602ffc7065618bcc3b87c46eaeb15a057084

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/12803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/1934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/148745 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/93508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/142294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/12957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/107643 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/93508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/143372 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/10404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/125700 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/88542 "Passed tests") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/139766 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/10023 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/7578 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/8854 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/119262 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/1716 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/151371 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/12491 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/1785 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/115949 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12506 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/10699 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/116287 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/11466 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/122189 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/67509 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21676 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12533 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/1644 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/12273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/76233 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/12471 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/12317 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/DragActions.serialization.in:
```diff
@@ -85,3 +85,10 @@ enum class WebCore::DragSourceAction : uint8_t {
+    bool handled;
+    std::optional<WebCore::FrameIdentifier> targetFrameID;
+    std::optional<WebCore::RemoteUserInputEventData> remoteUserInputEventData;
+}
```
This should be in `WebCoreArgumentCoders.serialization.in`, like `DragData`.

pxlcoder commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +238,90 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
-    if (!localMainFrame)
-        return false;
+    IntPoint point = frame.protectedView()->windowToContents(dragData.clientPosition());
+    HitTestResult hitTestResult = HitTestResult(point);
```
```suggestion
    auto hitTestResult = HitTestResult { point };
```

pxlcoder commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -141,7 +141,6 @@ bool isDraggableLink(const Element& element)
 }
 
 #if ENABLE(DRAG_SUPPORT)
-    
```
Unexpected change?

pxlcoder commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +238,90 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
         shouldOpenExternalURLsPolicy = document->shouldOpenExternalURLsPolicyToPropagate();
 
-    if ((m_dragDestinationActionMask.contains(DragDestinationAction::DHTML)) && dragIsHandledByDocument(m_dragHandlingMethod)) {
+    RefPtr<RemoteFrame> remoteFrame(nullptr);
```
Defaults to `null`.
```suggestion
    RefPtr<RemoteFrame> remoteFrame;
```

pxlcoder commented on Source/WebCore/page/DragController.h:
```diff
@@ -56,9 +57,14 @@ class PlatformMouseEvent;
-class DragController {
+struct DragDataInfo {
+    bool handled;
+    std::optional<FrameIdentifier> targetFrameID;
```
`remoteUserInputEventData` already has a `targetFrameID`. Is there ever a case where the two will be different?

jesxilin replied:
remoteUserInputEventData is populated / created in EventHandler. In the case that remoteFrame is found in DragController, we return `remoteFrame->frameID()` as the `targetFrameID` without calling EventHandler code.

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -41,6 +41,11 @@
+#if ENABLE(DRAG_SUPPORT)
+#include "SandboxExtension.h"
+#include <WebCore/DragData.h>
+#endif
```
Should be forward declared. See the comment at the top of the file.

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -59,7 +59,9 @@
 #if ENABLE(DRAG_SUPPORT)
+#include "SandboxExtension.h"
 #include <WebCore/DragActions.h>
+#include <WebCore/DragData.h>
```
I don't think the include is necessary here? Add it in implementation files if necessary.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -82,6 +82,12 @@
+#include "WebPageProxy.h"
+#include <WebCore/DragController.h>
+#include <WebCore/DragData.h>
+#endif
```
Same comment about forward declarations.

pxlcoder commented on Source/WebKitLegacy/mac/WebCoreSupport/WebDragClient.mm:
```diff
@@ -129,7 +129,6 @@ static WebDragDestinationAction kit(WebCore::DragDestinationAction action)
     [[m_webView _UIDelegateForwarder] webView:m_webView willPerformDragDestinationAction:kit(action) forDraggingInfo:dragData.platformData()];
 }
 
-
```
Unnecessary change in this PR, but also okay if you really want to...

achristensen07 replied:
nit: if we're not changing anything else in this file, let's omit this newline deletion.

pxlcoder commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2014,7 +2014,14 @@ - (BOOL)_tryToPerformDataInteraction:(id <UIDropSession>)session client:(CGPoint
+    RefPtr frame = page->focusController().focusedOrMainFrame();
+    if (!frame)
+        return false;
+    WebCore::DragDataInfo dragDataInfo = _private->page->dragController().performDragOperation(WTF::move(dragData), *frame);
```
```suggestion
    auto dragDataInfo = _private->page->dragController().performDragOperation(WTF::move(dragData), *frame);
```

pxlcoder commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -6403,10 +6411,10 @@ - (BOOL)performDragOperation:(id <NSDraggingInfo>)draggingInfo
     }
-    bool returnValue = core(self)->dragController().performDragOperation(WebCore::DragData { *dragData });
+    RefPtr localMainFrame = dynamicDowncast<WebCore::LocalFrame>(_private->page->mainFrame());
+    WebCore::DragDataInfo dragDataInfo = core(self)->dragController().performDragOperation(WebCore::DragData { *dragData }, *localMainFrame);
```
```suggestion
    auto dragDataInfo = core(self)->dragController().performDragOperation(WebCore::DragData { *dragData }, *localMainFrame);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -2218,6 +2218,111 @@ HTTPServer server({
+    RetainPtr simulator = adoptNS([[DragAndDropSimulator alloc] initWithWebViewFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
+
+    RetainPtr webView = [simulator webView];
+    webView.get().navigationDelegate = navigationDelegate.get();
```
```suggestion
    [webView setNavigationDelegate:navigationDelegate.get()];
```

pxlcoder commented on Source/WebCore/page/DragController.h:
```diff
@@ -56,9 +57,14 @@ class PlatformMouseEvent;
 
-class DragController {
+struct DragDataInfo {
+    bool handled;
```
```suggestion
    bool handled { false };
```

pxlcoder commented on Source/WebCore/page/DragController.h:
```diff
@@ -56,9 +57,14 @@ class PlatformMouseEvent;
-struct RemoteUserInputEventData;
 
-class DragController {
+struct DragDataInfo {
```
I think this should be in its own file, like `DragData`. Avoids needing to import `DragController.h` just to use `DragDataInfo.h`.

jesxilin replied:
Gotcha, will do!

pxlcoder commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +238,90 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
-bool DragController::performDragOperation(DragData&& dragData)
+DragDataInfo DragController::performDragOperation(DragData&& dragData, LocalFrame& frame)
 {
+    DragDataInfo dragDataInfo(false, std::nullopt, std::nullopt);
```
Default initialized.
```suggestion
    DragDataInfo dragDataInfo;
```

pxlcoder commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -137,7 +138,7 @@ extern const unsigned InvalidTouchIdentifier;
 enum CheckDragHysteresis { ShouldCheckDragHysteresis, DontCheckDragHysteresis };
 
-class EventHandler final : public CanMakeCheckedPtr<EventHandler> {
+class EventHandler final : public CanMakeCheckedPtr<EventHandler>, public CanMakeWeakPtr<EventHandler> {
```
Is this change necessary? Didn't spot usage of it with a quick scan.

pascoej commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3852,6 +3853,32 @@ void WebPageProxy::dragExited(DragData& dragData)
+#if PLATFORM(COCOA)
+void WebPageProxy::propagateDragAndDrop(FrameIdentifier frameID, DragData&& dragData, std::optional<RemoteUserInputEventData> remoteUserInputEventData, std::optional<SandboxExtension::Handle>&& sandboxExtensionHandle, std::optional<Vector<SandboxExtension::Handle>>&& sandboxExtensionsForUpload)
+{
+    DragData dragDataCopy(dragData);
```
Why do we need to copy `dragData`?

jesxilin replied:
I need to move it in `Messages::WebPage::PerformDragOperation` and also use it within the lambda

pascoej commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3866,16 +3893,29 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
+        if (!protectedThis)
+            return;
+
+        DragData dragDataCopy(dragData);
```
"

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3851,6 +3853,32 @@ void WebPageProxy::dragExited(DragData& dragData)
+            return;
+        }
+
+        protectedThis->protectedPageClient()->didPerformDragOperation(dragEventTargetData->handled);
```
If we are about to propagate again below, why do we notify client here?

jesxilin replied:
Hmm, you're right, we should notify the client _if_ we don't propagate again. I will change the ordering! TY

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3865,16 +3893,29 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
+            if (!dragEventTargetData)
+                protectedThis->protectedPageClient()->didPerformDragOperation(false);
+            else {
+                protectedThis->protectedPageClient()->didPerformDragOperation(dragEventTargetData->handled);
```
Same question: if we are about to propagate again below, why do we notify client here?

jesxilin replied:
Will change!

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3865,16 +3893,29 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
+
+        protectedThis->sendWithAsyncReplyToProcessContainingFrame(protectedThis->m_mainFrame->frameID(), Messages::WebPage::PerformDragOperation(protectedThis->m_mainFrame->frameID(), WTF::move(dragData), WTF::move(sandboxExtensionHandle), WTF::move(sandboxExtensionsForUpload)), [protectedThis, frameID = protectedThis->m_mainFrame->frameID(), dragDataCopy = WTF::move(dragDataCopy)] (std::optional ...
+            if (!dragEventTargetData)
+                protectedThis->protectedPageClient()->didPerformDragOperation(false);
```
To avoid indentation, we might just early return here

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3865,16 +3893,29 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
+        if (!dragEventTargetData)
+            protectedThis->protectedPageClient()->didPerformDragOperation(false);
+        else
+            protectedThis->protectedPageClient()->didPerformDragOperation(dragEventTargetData->handled);
```
maybe just `bool handled = dragEventTargetData ? dragEventTargetData->handled : false`?

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5530,14 +5542,23 @@ void WebPage::dragEnded(std::optional<FrameIdentifier> frameID, IntPoint clientP
-    m_sandboxExtensionTracker.willPerformLoadDragDestinationAction(WTF::move(m_pendingDropSandboxExtension));
+    if (m_pendingDropSandboxExtensionHandle) {
+        m_sandboxExtensionTracker.willPerformLoadDragDestinationAction(SandboxExtension::create(WTF::move(m_pendingDropSandboxExtensionHandle.value())));
+        m_pendingDropSandboxExtensionHandle = std::nullopt;
```
we could `if (auto pendingDropSandboxExtensionHandle = std::exchange(m_pendingDropSandboxExtensionHandle, std::nullopt)) { }`

jesxilin replied:
Ooh, gotcha!

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5530,14 +5542,23 @@ void WebPage::dragEnded(std::optional<FrameIdentifier> frameID, IntPoint clientP
+            extension->consumePermanently();
+    }
+
+    m_pendingDropExtensionHandlesForFileUpload->clear();
```
Why not set to `std::nullopt`?

jesxilin replied:
No reason, can change :)

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5480,24 +5481,35 @@ void WebPage::performDragControllerAction(std::optional<FrameIdentifier> frameID
+
+    if (dragEventTargetData.handled) {
+        m_pendingDropSandboxExtensionHandle = std::nullopt;
+        m_pendingDropExtensionHandlesForFileUpload->clear();
```
Why not set to `std::nullopt`?

jesxilin replied:
No particular reason, I just followed the original code but since it is an `std::optional` now, I can definitely do that instead!

achristensen07 commented on Source/WebCore/platform/DragEventTargetData.h:
```diff
@@ -0,0 +1,38 @@
+struct DragEventTargetData {
+    bool handled { false };
+    std::optional<FrameIdentifier> targetFrameID;
+    std::optional<RemoteUserInputEventData> remoteUserInputEventData;
```
It may make it a little more clear to use something like this to indicate we either have a local result or data to forward to another process:

```
struct ForwardingInfo {
    FrameIdentifier targetFrameID;
    RemoteUserInputEventData remoteUserInputData;
};
enum class Handled : bool { Yes, No };
Variant<Handled, ForwardingInfo> data;
```

That expresses that we either have one or the other, and if we have forwarding info we always have both members.

jesxilin replied:
I like this structure too, it does seem clearer. I'm working on it~

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "WebPage.h"
 #include "WebPageProxyMessages.h"
 #include <WebCore/AXObjectCache.h>
+#include <WebCore/DragData.h>
```
It's probably worth doing all the unified build fixes and Frame/LocalFrame changes in a separate PR that clearly doesn't change behavior.  That way it'll be easier to isolate any issues to a smaller PR.

jesxilin replied:
Ooh, ok!

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,94 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
         clearDragCaret();
-        return true;
+        dragEventTargetData = DragEventHandled::Yes;
+        return dragEventTargetData;
```
Can this be just return { DragEventHandled::Yes }?

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,94 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
-bool DragController::performDragOperation(DragData&& dragData)
+DragEventTargetData DragController::performDragOperation(DragData&& dragData, LocalFrame& frame)
 {
+    DragEventTargetData dragEventTargetData;
```
Can we remove this?  It's not ideal to have a default-constructed object that should never be used in its default-constructed state.

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,94 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
+
+    if (remoteFrame) {
+        dragEventTargetData = remoteFrame->frameID();
+        return dragEventTargetData;
```
return { remoteFrame->frameID() };

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,94 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
             clearDragCaret();
             m_documentUnderMouse = nullptr;
-            return true;
+            return dragEventTargetData;
```
return { DragEventHandled::Yes };

If the return type of performDragAndDrop is changed it will be much more clear how this code flows.

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,94 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
         clearDragCaret();
-        return true;
+        dragEventTargetData = DragEventHandled::Yes;
+        return dragEventTargetData;
```
return { DragEventHandled::Yes };
etc.  Same below.

achristensen07 commented on Source/WebCore/page/DragController.h:
```diff
@@ -27,9 +27,11 @@
 
 #include <WebCore/Document.h>
 #include <WebCore/DragActions.h>
+#include <WebCore/DragEventTargetData.h>
```
I think both of these new includes can instead be just forward declarations.

achristensen07 commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2869,14 +2871,32 @@ void EventHandler::cancelDragAndDrop(const PlatformMouseEvent& event, std::uniqu
+    if (RefPtr remoteFrame = dynamicDowncast<RemoteFrame>(subframe)) {
+        if (auto remoteUserInputEventData = userInputEventDataForRemoteFrame(remoteFrame.get(), result.roundedPointInInnerNodeFrame())) {
+            dragEventTargetData = remoteUserInputEventData->targetFrameID;
+            return false;
```
I think it would be better if this function returned a DragEventTargetData instead of taking an out-parameter, which usually leads to convoluted code.

achristensen07 commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -27,6 +27,7 @@
 
 #include <WebCore/Cursor.h>
 #include <WebCore/DragActions.h>
+#include <WebCore/DragEventTargetData.h>
```
I think this should be just a forward declaration of the return type instead of including a header from another header, which slows down the build.

achristensen07 commented on Source/WebCore/platform/DragEventTargetData.h:
```diff
@@ -0,0 +1,51 @@
+
+namespace WTF {
+
+template<> inline constexpr bool isValidEnum<WebCore::DragEventHandled>(std::underlying_type_t<WebCore::DragEventHandled> value)
```
This should be replaced by an entry in a *.serialization.in file.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3851,6 +3853,32 @@ void WebPageProxy::dragExited(DragData& dragData)
+{
+    DragData dragDataCopy(dragData);
+
+    grantAccessToCurrentPasteboardData(dragDataCopy.pasteboardName(), [] () { }, frameID);
```
Will this cause a race condition?  Should we move the sendWithAsyncReplyToProcessContainingFrame call to the lambda here?

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5480,24 +5481,35 @@ void WebPage::performDragControllerAction(std::optional<FrameIdentifier> frameID
 }
 
-void WebPage::performDragOperation(WebCore::DragData&& dragData, SandboxExtension::Handle&& sandboxExtensionHandle, Vector<SandboxExtension::Handle>&& sandboxExtensionsHandleArray, CompletionHandler<void(bool)>&& completionHandler)
+void WebPage::performDragOperation(std::optional<WebCore::FrameIdentifier> frameID, WebCore::DragData&& dragData, std::optional<SandboxExtension::Handle>&& sandboxExtensionHandle, std::optional<Vector<SandboxExtension::Handle>>&& sandboxExtensionsForUpload, CompletionHandler<void(std::optional<WebCore::DragEventTargetData>, std::optional<SandboxExtension::Handle>&&, std::optional<Vector<SandboxEx ...
```
I think DragEventTargetData shouldn't be optional.  We had a non-optional bool before, and we are just adding one new case to it.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5480,24 +5481,35 @@ void WebPage::performDragControllerAction(std::optional<FrameIdentifier> frameID
+
+    DragEventTargetData dragEventTargetData = m_page->dragController().performDragOperation(WTF::move(dragData), *localFrame);
+
+    if (auto handled = std::holds_alternative<WebCore::DragEventHandled>(dragEventTargetData); handled && (std::get<WebCore::DragEventHandled>(dragEventTargetData) == WebCore::DragEventHandled::Yes)) {
```
std::get_if seems helpful here.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -35,6 +35,7 @@
 #include <WebCore/DictionaryPopupInfo.h>
 #include <WebCore/DisabledAdaptations.h>
 #include <WebCore/DragActions.h>
+#include <WebCore/DragEventTargetData.h>
```
forward declaration

achristensen07 commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2014,7 +2015,16 @@ - (BOOL)_tryToPerformDataInteraction:(id <UIDropSession>)session client:(CGPoint
+        return false;
+    auto dragEventTargetData = _private->page->dragController().performDragOperation(WTF::move(dragData), *frame);
+    if (std::holds_alternative<WebCore::DragEventHandled>(dragEventTargetData))
+        return static_cast<bool>(std::get<WebCore::DragEventHandled>(dragEventTargetData));
```
This should probably use a ternary operator and comparison to WebCore::DragEventHandled::Yes rather than casting to a bool, which assumes WebCore::DragEventHandled::No is the first value of the enum.  Also, std::get_if

achristensen07 commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2014,7 +2015,16 @@ - (BOOL)_tryToPerformDataInteraction:(id <UIDropSession>)session client:(CGPoint
+    auto dragEventTargetData = _private->page->dragController().performDragOperation(WTF::move(dragData), *frame);
+    if (std::holds_alternative<WebCore::DragEventHandled>(dragEventTargetData))
+        return static_cast<bool>(std::get<WebCore::DragEventHandled>(dragEventTargetData));
+    return true;
```
This should never happen, but I think return false would probably be better here.

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,83 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
-    RefPtr focusedOrMainFrame = m_page->focusController().focusedOrMainFrame();
-    if (!focusedOrMainFrame)
-        return false;
+    RefPtr ignoreSelectionChangesFrame = &frame;
```
This could be a Ref, it could also be removed.

achristensen07 commented on Source/WebCore/page/DragController.h:
```diff
@@ -30,6 +30,7 @@
 #include <WebCore/DragImage.h>
 #include <WebCore/IntPoint.h>
 #include <WebCore/IntRect.h>
+#include <WebCore/RemoteUserInputEventData.h>
```
This include seems unnecessary to have in a header.  Can the forward declaration work?

achristensen07 commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2869,21 +2871,35 @@ void EventHandler::cancelDragAndDrop(const PlatformMouseEvent& event, std::uniqu
-            preventedDefault = targetFrame->eventHandler().performDragAndDrop(event, WTF::move(pasteboard), sourceOperationMask, draggingFiles);
+        if (RefPtr localTargetFrame = dynamicDowncast<LocalFrame>(targetFrame)) {
+            auto dragEventTargetData = localTargetFrame->eventHandler().performDragAndDrop(event, WTF::move(pasteboard), sourceOperationMask, draggingFiles, result, WTF::move(dragData));
+            return dragEventTargetData;
```
These paths are now missing the calls to clearDragState, which is probably not intended.

jesxilin replied:
Good catch! Thanks for lmk!

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3851,6 +3853,28 @@ void WebPageProxy::dragExited(DragData& dragData)
+        protectedThis->sendWithAsyncReplyToProcessContainingFrame(frameID, Messages::WebPage::PerformDragOperation(frameID, WTF::move(dragData), WTF::move(sandboxExtensionHandle), WTF::move(sandboxExtensionsForUpload)), [protectedThis, frameID, dragDataCopy = WTF::move(dragDataCopy), dragStorageName] (WebCore::DragEventTargetData dragEventTargetData, std::optional<SandboxExtension::Handle>&& sand ...
+            WTF::switchOn(dragEventTargetData, [&protectedThis](WebCore::DragEventHandled handled) {
+                protectedThis->protectedPageClient()->didPerformDragOperation(static_cast<bool>(handled));
+            }, [&protectedThis, frameID, &dragDataCopy, &sandboxExtensionHandle, &sandboxExtensionsForUpload, dragStorageName](FrameIdentifier targetFrameID) mutable {
```
The switchOn implementation has NOESCAPE, so we can safely just capture everything by reference with [&]

jesxilin replied:
Ooh gotcha!

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5530,14 +5542,21 @@ void WebPage::dragEnded(std::optional<FrameIdentifier> frameID, IntPoint clientP
+    if (!m_pendingDropExtensionHandlesForFileUpload)
+        return;
+
+    for (size_t i = 0; i < m_pendingDropExtensionHandlesForFileUpload->size(); i++) {
```
We could keep the C++11-style iteration by just checking if (!m_pendingDropExtensionsForFileUpload) return; then iterating *m_pendingDropExtensionsForFileUpload instead and calling m_pendingDropExtensionsForFileUpload->clear()

jesxilin replied:
I had `m_pendingDropExtensionsForFileUpload->clear()` but was told I could just do `m_pendingDropExtensionHandlesForFileUpload = std::nullopt;` since it's an optional. Which is better?

achristensen07 commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -6403,10 +6414,12 @@ - (BOOL)performDragOperation:(id <NSDraggingInfo>)draggingInfo
     delete dragData;
-
-    return returnValue;
+    if (auto handled = std::holds_alternative<WebCore::DragEventHandled>(dragEventTargetData); handled && ( std::get<WebCore::DragEventHandled>(dragEventTargetData) == WebCore::DragEventHandled::Yes))
```
nit: extra space after (

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,83 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
+    if (RefPtr subframe = EventHandler::subframeForTargetNode(hitTestResult.protectedTargetNode().get()))
+        remoteFrame = dynamicDowncast<RemoteFrame>(subframe);
+
+    if (remoteFrame)
```
nit: remoteFrame is only used once and is only non-null in one case.  I think this would be a bit more concise:

if (RefPtr remoteFrame = dynamicDowncast<RemoteFrame>(EventHandler::subframeForTargetNode(hitTestResult.protectedTargetNode().get())))
    return { remoteFrame->frameID() };

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,83 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
 }
 
-bool DragController::performDragOperation(DragData&& dragData)
+DragEventTargetData DragController::performDragOperation(DragData&& dragData, LocalFrame& frame)
```
Can the LocalFrame parameter be const?

jesxilin replied:
I could change it but then I'd have to also change `FrameLoadRequest.h`

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,83 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
+    auto hitTestResult = HitTestResult { point };
 
-    m_documentUnderMouse = localMainFrame->documentAtPoint(dragData.clientPosition());
+    if (frame.contentRenderer()) {
```
This seems to duplicate code from LocalFrame::documentAtPoint a bit.  Why can't we just use LocalFrame::documentAtPoint?  If we can't, I think it would be good to still share the code somewhere by refactoring instead of duplicating.

jesxilin replied:
It's b/c I use `result` in `LocalFrame::documentAtPoint` aka `hitTestResult` in other places of the `DragController` code. I could refactor `LocalFrame::documentAtPoint` maybe to return a HitTestResult or alter a HitTestResult as an arg but then it should prob be renamed since it's not rly "document at point" anymore. Or I could do `m_documentUnderMouse = frame.documentAtPoint(dragData.clientPosition());` and then call `frame.eventHandler().hitTestResultAtPoint` again separately but that's also redundant. Wdyt?

hortont424 replied:
I think Alex is suggesting e.g. moving the parts of "documentAtPoint" that both "documentAtPoint" and you need out into a new method ("hitTestAtPoint"? do we really not already have such a thing?) and having both "documentAtPoint" and DragController call the factored-out method.

jesxilin replied:
Ahhh, I understand now. There is a `hitTestResultAtPoint` in `EventHandler` that `documentAtPoint` uses. Perhaps a function that returns HitTestResult, which then `documentAtPoint` can use and DragController can use...

jesxilin replied:
This is definitely doable, but I don't feel it belongs in this PR so I will make a separate followup that makes this change. :)

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,83 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
+
+        DragEventTargetData performDragAndDropResult = frame.eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles(), hitTestResult, WTF::move(dragData));
+
+        if (auto handled = std::get_if<DragEventHandled>(&performDragAndDropResult); *handled == DragEventHandled::Yes) {
```
This will dereference null without checking if performDragAndDropResult doesn't contain a DragEventHandled.
You need this:
if (auto handled = std::get_if<DragEventHandled>(&performDragAndDropResult); handled && *handled == DragEventHandled::Yes)

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5480,24 +5481,35 @@ void WebPage::performDragControllerAction(std::optional<FrameIdentifier> frameID
+    DragEventTargetData dragEventTargetData = m_page->dragController().performDragOperation(WTF::move(dragData), *localFrame);
+
+    if (auto handled = std::get_if<WebCore::DragEventHandled>(&dragEventTargetData); *handled == WebCore::DragEventHandled::Yes) {
+        m_pendingDropSandboxExtensionHandle = std::nullopt;
```
This doesn't seem quite right to me.  I think the sandbox extensions should be consumed instead of just discarding the handles, right?  Maybe we could assert that something else has already consumed them.

jesxilin replied:
Yeah, if it satisfies the `if` condition, it _should_ have been consumed if it exists. I can try to assert that it is consumed/null or smth along those lines :)

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -361,7 +361,7 @@ messages -> WebPage WantsAsyncDispatchMessage {
 #if !PLATFORM(GTK) && ENABLE(DRAG_SUPPORT)
     PerformDragControllerAction(std::optional<WebCore::FrameIdentifier> frameID, enum:uint8_t WebKit::DragControllerAction action, WebCore::DragData dragData) -> (enum:uint8_t std::optional<WebCore::DragOperation> dragOperation, enum:uint8_t WebCore::DragHandlingMethod dragHandlingMethod, bool mouseIsOverFileInput, unsigned numberOfItemsToBeAccepted, WebCore::IntRect insertionRect, WebCore::IntRe ...
-    PerformDragOperation(WebCore::DragData dragData, WebKit::SandboxExtensionHandle sandboxExtensionHandle, Vector<WebKit::SandboxExtensionHandle> sandboxExtensionsForUpload) -> (bool handled)
+    PerformDragOperation(std::optional<WebCore::FrameIdentifier> frameID, WebCore::DragData dragData, std::optional<WebKit::SandboxExtensionHandle> sandboxExtensionHandle, std::optional<Vector<WebKit::SandboxExtensionHandle>> sandboxExtensionsForUpload) -> (WebCore::DragEventTargetData dragEventTargetData, std::optional<WebKit::SandboxExtensionHandle> sandboxExtensionHandle, std::optional<Vector< ...
```
There shouldn't be a need to make the sandboxExtensionHandle  and sandboxExtensionsForUpload going to the web process optional, and the result should only have the handles if the result has a frame identifier.  I think this should look like this:

PerformDragOperation(std::optional<WebCore::FrameIdentifier> frameID, WebCore::DragData dragData, WebKit::SandboxExtensionHandle sandboxExtensionHandle, Vector<WebKit::SandboxExtensionHandle> sandboxExtensionsForUpload) -> (WebKit::DragOperationResult result)

DragOperationResult then looks something like this:

struct DragOperationForwardingInfo {
    FrameIdentifier frameID;
    SandboxExtensionHandle sandboxExtensionHandle;
    Vector<WebKit::SandboxExtensionHandle> sandboxExtensionsForUpload;
};
using DragOperationResult  = Expected<DragEventHandled, DragOperationForwardingInfo>;

pxlcoder commented on Source/WebCore/page/EventHandler.cpp:
It looks like you call `clearDragState()` in every place before this method returns. It would be good to use `makeScopeExit` at the top of the method instead, and remove the individual calls to `clearDragState()`. 

```
auto scopeExit = makeScopeExit([&] {
    clearDragState();
});
```

(unless I'm missing some codepath that does not call `clearDragState()` before returning)

jesxilin replied:
I believe you are right! All of the cases should be calling `clearDragState()`, so I will utilize what you suggested. Thanks!

pxlcoder commented on Source/WebCore/page/EventHandler.h:
Can this be forward-declared?

jesxilin replied:
Yes, will do!

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5480,24 +5482,41 @@ void WebPage::performDragControllerAction(std::optional<FrameIdentifier> frameID
+            };
+            completionHandler(WTF::move(result));
+            return;
+        }
```
This case needs to call the completion handler

jesxilin replied:
Will do!

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,75 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
+
+        DragEventTargetData performDragAndDropResult = frame.eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles(), hitTestResult, WTF::move(dragData));
+
+        if (auto handled = std::get_if<DragEventHandled>(&performDragAndDropResult);  handled && *handled == DragEventHandled::Yes) {
```
nit: extra space after semicolon.

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,75 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
 
-    m_documentUnderMouse = localMainFrame->documentAtPoint(dragData.clientPosition());
+    if (frame.contentRenderer()) {
+        static constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowChildFrameContent };
```
Can this call LocalFrame::documentAtPoint or at least share some code rather than duplicating this?  This leads to future updates in one location that will forget to update this duplicate copy.

jesxilin replied:
I think this is what you said here: https://github.com/WebKit/WebKit/pull/53785#discussion_r2692300222? I will be making a follow-up pr right after this to add a new function (that extracts the duplicate code they share) that `documentAtPoint` and `dragController` can both use. I only didn't do it in this pr I felt like it doesn't belong here and should be a separate refactoring pr.

achristensen07 commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -239,74 +240,75 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
-            preventedDefault = localMainFrame->eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles());
-        if (preventedDefault) {
+
+        DragEventTargetData performDragAndDropResult = frame.eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles(), hitTestResult, WTF::move(dragData));
```
The use of dragData needs to be on the previous lines.  Otherwise, drag data may or may not have been moved from when it is passed to createMouseEvent depending on undefined behavior.  Something like this:
```
auto mouseEvent = createMouseEvent(dragData);
auto pasteboard = Pasteboard::create(dragData);
auto mask = dragData.draggingSourceOperationMask();
auto containsFiles = dragData.containsFiles();
DragEventTargetData performDragAndDropResult = frame.eventHandler().performDragAndDrop(WTF::move(mouseEvent), WTF::move(pasteboard), mask, containsFiles, hitTestResult, WTF::move(dragData));
```

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3865,16 +3891,25 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
+
+        DragData dragDataCopy(dragData);
+
+        protectedThis->sendWithAsyncReplyToProcessContainingFrame(protectedThis->m_mainFrame->frameID(), Messages::WebPage::PerformDragOperation(protectedThis->m_mainFrame->frameID(), WTF::move(dragData), WTF::move(sandboxExtensionHandle), WTF::move(sandboxExtensionsForUpload)), [protectedThis, frameID = protectedThis->m_mainFrame->frameID(), dragDataCopy = WTF::move(dragDataCopy), dragStorageNam ...
```
We should probably add an early return if m_mainFrame is null.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3865,16 +3891,25 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
-    sendWithAsyncReply(Messages::WebPage::PerformDragOperation(dragData, WTF::move(sandboxExtensionHandle), WTF::move(sandboxExtensionsForUpload)), [this, protectedThis = Ref { *this }] (bool handled) {
-        if (RefPtr pageClient = this->pageClient())
-            pageClient->didPerformDragOperation(handled);
+    sendWithAsyncReplyToProcessContainingFrame(m_mainFrame->frameID(), Messages::WebPage::PerformDragOperation(m_mainFrame->frameID(), WTF::move(dragData), WTF::move(sandboxExtensionHandle), WTF::move(sandboxExtensionsForUpload)), [protectedThis = Ref { *this }, frameID = m_mainFrame->frameID()] (DragOperationResult dragOperationResult) mutable {
```
Ditto, null check m_mainFrame.  It can be null.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3865,16 +3891,25 @@ void WebPageProxy::performDragOperation(DragData& dragData, const String& dragSt
-        if (RefPtr pageClient = this->pageClient())
-            pageClient->didPerformDragOperation(handled);
+    sendWithAsyncReplyToProcessContainingFrame(m_mainFrame->frameID(), Messages::WebPage::PerformDragOperation(m_mainFrame->frameID(), WTF::move(dragData), WTF::move(sandboxExtensionHandle), WTF::move(sandboxExtensionsForUpload)), [protectedThis = Ref { *this }, frameID = m_mainFrame->frameID()] (DragOperationResult dragOperationResult) mutable {
+        protectedThis->protectedPageClient()->didPerformDragOperation(std::get_if<bool>(&dragOperationResult));
```
This is incorrect.  std::get_if returns a pointer and if it's null it'll call didPerformDragOperation with false and if it's non-null it'll call with true.  I think you want this instead:
protectedThis->protectedPageClient()->didPerformDragOperation(std::holds_alternative<bool>(dragOperationResult) ? std::get<bool>(dragOperationResult) : false)

jesxilin replied:
Yes, thank you!

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
You don't need a separate `shouldReturn` variable. `CompletionHandler` nulls out its `m_function` when it's called. And the boolean operator checks the existence of `m_function`.

Which means this could just be:

```
if (completionHandler)
    completionHandler(false);
```

pxlcoder commented on Source/WebCore/page/EventHandler.cpp:
Nit: The local variable is unnecessary. Could just be `return ...`.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```suggestion
    auto dragEventTargetData = m_page->dragController().performDragOperation(WTF::move(dragData), *localFrame);
```

jesxilin committed:
[Site Isolation] Drag & Drop within remote frame is nonfunctional
https://bugs.webkit.org/show_bug.cgi?id=301963
rdar://160806781

Reviewed by Aditya Keerthi and Alex Christensen.

Original patch by pascoej. Thanks to achristensen07 for suggesting this
new approach.

In certain site isolation cases, dragging and dropping on macOS/iOS causes
a navigation when dropping items instead of a drop event.
The default mode for drag and drop is to load, which would then
trigger a navigation. When dropping with the presence of remote frames,
there is no logic to support this functionality and the behavior defaults
to load. To fix this unexpected behavior, implement support
for drag and drop with remote frames. This includes finding the correct
destination frame, and propagating the drag and drop action to the target
frame.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add new files that supports DragEventTargetData and DragEventForwardingData.
The DragEventForwardingData struct contains the information for propagating the event
to the target frame.

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::performDragOperation):
* Source/WebCore/page/DragController.h:
Alter the original code to factor in the presence of remote frames.
If a remote frame (not the current frame) is the destination of the
drop event, return a DragEventTargetData that represents the remote frame ID. This
remote frame ID will be used as the new destination frame ID. Otherwise,
handle cases for edit and load accordingly using the current frame instead
of the main frame.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::contentFrameForNode):
(WebCore::EventHandler::updateDragAndDrop):
(WebCore::EventHandler::cancelDragAndDrop):
(WebCore::EventHandler::performDragAndDrop):
* Source/WebCore/page/EventHandler.h:
The return value in EventHandler::performDragAndDrop indicates if a drop event
was successfully handled or not.

Check if the subframeForTargetNode is a remote frame or a local frame.
If the subframe is a remote frame, return the targetFrameID.

If the current frame is the owner of the target frame for the drop event,
there are two cases: local and remote frame. If the target frame is a local frame,
simply call performDragAndDrop on the target frame. In this case, the target frame
is accessible. If the target frame is remote, the target frame is not accessible
in this context. Thus, populate DragEventTargetData targetFrameID to be used to
propagate to the remote frame.

Otherwise, if it's at the target, create DataTransfer::createForDrop and
return whether the drop event was successfully executed.

* Source/WebCore/platform/DragEventTargetData.h: Added.
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/DragEventForwardingData.h: Added.
* Source/WebKit/Shared/DragEventForwardingData.serialization.in: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Add new files for the DragEventTargetData and DragOperationResult variants.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::propagateDragAndDrop):
(WebKit::WebPageProxy::performDragOperation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
DragOperationResult contains either a boolean that indicates the equivalent of
the DragEventHandled in DragEventTargetData (whether the drag/drop event was successful),
or a DragEventForwardingData struct that neatly packages all the relevant information needed
for the WebPageProxy to forward the drag event to the correct frame. DragEventForwardingData
consists of the targetFrameID and sandbox extension handles needed.

WebPageProxy::performDragOperation utilizes a new model for drag and drop to include remote frames.
Previously, the code only accounted for the local main frame.

This is the current model:

1) WebPageProxy's performDragOperation is responsible for sending the initial
IPC call to WebPage::PerformDragOperation, in the context of the main frame.

2) A reply (completion handler) is received from WebPage::PerformDragOperation.
If DragEventTargetData is of DragEventHandled type, it'll indicate if the event succeeded. In this
case, return a boolean in the completion handler.
If the type is a Frame Identifier, that means there is a frame that needs to receive this
operation and hasn't yet. So, create a DragEventForwardingData that includes all the information needed
for the operation and return it in the completion handler.

3) In propagateDragAndDrop, send an IPC message to WebPage::PerformDragOperation
with the DragEventForwardingData's targetFrameID. If it is still not handled and there is another Frame
Identifier that is not the current frame ID passed in, a propagation must happen again.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::performDragControllerAction):
(WebKit::WebPage::performDragOperation):
(WebKit::WebPage::willPerformLoadDragDestinationAction):
(WebKit::WebPage::mayPerformUploadDragDestinationAction):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
DragEventTargetData used in WebPage consists of a DragEventHandled that represents if the drag event
was successful or not, and a Frame Identifier that represents the frameID of the
target. Also, change the sandboxing logic to use SandboxExtension handles instead. This is because
when WebPage::performDragOperation is called via IPC messages, now
there is no guaranteee that the extensions will be used in the current frame context.
As a result, there needs to be a way to retain the SandboxExtension handles, which
will later be utilized to create the extensions to be consumed for dropping, on the
actual target frame.

* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _tryToPerformDataInteraction:client:global:operation:]):
(-[WebView performDragOperation:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, DragAndDropWithoutNavigation)):
Create an API test to ensure a navigation does not occur when dragging and dropping.

Canonical link: https://commits.webkit.org/306143@main
